src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs

```diff
@@ -387,40 +387,58 @@ private static void CreateParentsAndDirectory(string fullPath)
             }
         }
 
-        public static void MoveDirectory(string sourceFullPath, string destFullPath)
+        private static void MoveDirectory(string sourceFullPath, string destFullPath, bool sameDirectoryDifferentCase)
         {
-            // Windows doesn't care if you try and copy a file via "MoveDirectory"...
-            if (FileExists(sourceFullPath))
-            {
-                // ... but it doesn't like the source to have a trailing slash ...
+            ReadOnlySpan<char> srcNoDirectorySeparator = Path.TrimEndingDirectorySeparator(sourceFullPath.AsSpan());
+            ReadOnlySpan<char> destNoDirectorySeparator = Path.TrimEndingDirectorySeparator(destFullPath.AsSpan());
 
+            if (OperatingSystem.IsBrowser() && Path.EndsInDirectorySeparator(sourceFullPath) && FileExists(sourceFullPath))
+            {
                 // On Windows we end up with ERROR_INVALID_NAME, which is
                 // "The filename, directory name, or volume label syntax is incorrect."
-                //
-                // This surfaces as an IOException, if we let it go beyond here it would
-                // give DirectoryNotFound.
-
-                if (Path.EndsInDirectorySeparator(sourceFullPath))
-                    throw new IOException(SR.Format(SR.IO_PathNotFound_Path, sourceFullPath));
-
-                // ... but it doesn't care if the destination has a trailing separator.
-                destFullPath = Path.TrimEndingDirectorySeparator(destFullPath);
+                // On Unix, rename fails with ENOTDIR, but on WASM it does not.
+                // So if the path ends with directory separator, but it's a file, we just throw.
+                throw new IOException(SR.Format(SR.IO_PathNotFound_Path, sourceFullPath));
             }
 
-            if (FileExists(destFullPath))
+            if (!sameDirectoryDifferentCase) // This check is to allow renaming of directories
```

Comment:
Because file systems are case sensitive I don't understand why we need a specific path for `sameDirectoryDifferentCase`?

Reply:
this is something that already existed and to be honest I have not questioned that before you have asked the question. It seems that @wfurt agrees with you (https://github.com/dotnet/runtime/pull/65059#issue-1128165969):

> The MoveDirectory_FailToMoveLowerCaseDirectoryWhenUpperCaseDirectoryExists is technically wrong on Unix IMHO.
The case sensitivity is determined by given filesystem not by the OS itself.

Reply:
My guess is that this may be there to surface better exceptions. We _could_ defer to actual IO but that may give different behaviors on different platforms. And Some OSes and file systems do care about case and some don't.